unconstrained_test.go

```diff
@@ -663,55 +726,71 @@ func testLocal(t *testing.T, tests []unconstrainedTest, method Method) {
 			t.Errorf("error finding minimum (%v) for:\n%v", err, test)
 			continue
 		}
-
 		if result == nil {
 			t.Errorf("nil result without error for:\n%v", test)
 			continue
 		}
 
-		funcInfo := newFunctionInfo(test.f)
-
-		// Evaluate the norm of the gradient at the found optimum location.
-		optF := funcInfo.function.Func(result.X)
-		var optNorm float64
-		if funcInfo.IsGradient {
-			g := make([]float64, len(test.x))
-			funcInfo.gradient.Grad(result.X, g)
-			optNorm = floats.Norm(g, math.Inf(1))
-		}
+		f := newFunctionInfo(test.f)
 
 		// Check that the function value at the found optimum location is
-		// equal to result.F
+		// equal to result.F.
+		optF := test.f.Func(result.X)
 		if optF != result.F {
 			t.Errorf("Function value at the optimum location %v not equal to the returned value %v for:\n%v",
 				optF, result.F, test)
 		}
+		if result.Gradient != nil {
+			// Evaluate the norm of the gradient at the found optimum location.
+			g := make([]float64, len(test.x))
+			f.gradient.Grad(result.X, g)
 
-		// Check that the norm of the gradient at the found optimum location is
-		// smaller than the tolerance.
-		if optNorm >= settings.GradientThreshold {
-			t.Errorf("Norm of the gradient at the optimum location %v not smaller than tolerance %v for:\n%v",
-				optNorm, settings.GradientThreshold, test)
+			if !floats.Equal(result.Gradient, g) {
+				t.Errorf("Gradient at the optimum location not equal to the returned value for:\n%v", test)
+			}
+
+			optNorm := floats.Norm(g, math.Inf(1))
+			// Check that the norm of the gradient at the found optimum location is
+			// smaller than the tolerance.
+			if optNorm >= settings.GradientThreshold {
+				t.Errorf("Norm of the gradient at the optimum location %v not smaller than tolerance %v for:\n%v",
+					optNorm, settings.GradientThreshold, test)
+			}
 		}
 
-		// We are going to restart the solution using a fixed starting gradient
-		// and value, so evaluate them.
+		if method == nil {
+			// The tests below make sense only if the method used is known.
+			continue
+		}
+
+		if !method.Needs().Gradient && !method.Needs().Hessian {
+			// At present, gradient-free methods can correctly terminate only
```

Comment:
This comment is not true. They can also converge with an error or if the maximum function evaluations is reached. The comment should state that "gradient-free tests can correctly terminate..."